Avoid LazilyDefinedAttributes being included multiple times in class, and remove methods after first use

shioyama opened the pull request:
Solution to issues in #35968 and #35966. Also fixes #34659.

### Summary

Currently, the acceptance validator includes an anonymous module in order to override `method_missing` and `respond_to_missing?` so that the validator can catch calls to attribute methods and define attribute accessors lazily (to avoid a schema load).

The problem with this is that each call to `validates` on a model will include another module into the class. I've already fixed a bunch of tests to avoid polluting test classes as a result of this, but the underlying problem is still there.

This PR gets around the issue using a normal module and grabbing the attributes from the validators on the model, which are accessible within method_missing and respond_to_missing?

@kamipo If this seems like a reasonable solution, I'd like to revert at least #35968 before merging.

shioyama committed:
Rewrite LazilyDefineAttributes so that it removes itself

LazilyDefineAttributes only needs to stay in the model's ancestors long
enough to define accessors for acceptance attributes, after that it is
not needed anymore. To optimize this, we eagerly define attributes as
soon as either +method_missing+ or +respond_to_missing?+ is called, then
immediately remove those methods from the module so that future calls
will incur no performance penalty.

I've also removed the call to +define_attribute_methods+ here, which
is no longer be necessary to avoid an infinite loop.

shioyama commented:
Also: `def` is faster than `define_method`, so this should have a small performance improvement for anything going through `method_missing` or `respond_to?`.

kamipo commented:
Reverting #35968 is less worth to me, but at least the comment should be removed, so either is fine.

https://github.com/rails/rails/blob/f8944ed22dc9453dea5814676fa49ce119a58726/activemodel/test/cases/validations/acceptance_validation_test.rb#L96-L98

And also, this fixes #34659 fortunately, can you add a test case for #34659?

kamipo commented on activemodel/lib/active_model/validations/acceptance.rb:
```diff
@@ -17,54 +17,54 @@ def validate_each(record, attribute, value)
+              def acceptance_attributes_for(klass)
+                klass.validators.
+                  grep(::ActiveModel::Validations::AcceptanceValidator).
+                  flat_map(&:attributes).map(&:to_s)
```
```suggestion
                  flat_map(&:attributes).map!(&:to_s)
```

shioyama commented:
> And also, this fixes #34659 fortunately, can you add a test case for #34659?

Oh wow, what a pleasant surprise! I'll add a test case later tonight. Thanks for quick response!

kamipo commented:
> Also: `def` is faster than `define_method`, so this should have a small performance improvement for anything going through `method_missing` or `respond_to?`, not to mention that if you've added acceptance validators multiple times, it will cut the number of overrides from N to just one.

I don't know this have a performance improvement, in exchange for `define_method` to `def`, this makes each `matches?` generates extra an array and strings.

shioyama commented:
cdb035c is an attempt to minimize the number of string conversions (partially). It seems that `attribute_method?` takes a symbol or string.

shioyama commented:
> I don't know this have a performance improvement, in exchange for define_method to def, this makes each matches? generates extra an array and strings.

Good point. `matches?` can actually stop once it finds a symbol that converts to a matching string, so I improved that a bit.

kamipo commented:
`record.respond_to?(:unknown_method)` is perhaps slower though.

shioyama commented:
Added a quick test in be91bd9.

kamipo commented on activemodel/lib/active_model/validations/acceptance.rb:
```diff
@@ -17,54 +17,54 @@ def validate_each(record, attribute, value)
 
         def setup!(klass)
-          klass.include(LazilyDefineAttributes.new(AttributeDefinition.new(attributes)))
+          klass.include LazilyDefineAttributes
```
How about memoizing `@acceptance_attributes` here?

```suggestion
          klass.include LazilyDefineAttributes
          @acceptance_attributes =
            (LazilyDefineAttributes.acceptance_attributes_for(klass) + attributes).map!(&:to_s)
```

shioyama replied:
Yeah, I know we could shove them somewhere in the class. But at that point, any application that happens to have defined `acceptance_attributes` (which is not unlikely) would potentially have some very strange bugs.

I don't really feel that there is such an issue here that we need to do something like this. At least for the "acceptance attributes", the first time *any* instance of a model hits method_missing, it will define the accessors and we should never hit `method_missing` again (for those attributes at least). So I think it's a relatively narrowly-scoped issue we're talking about.

shioyama replied:
I have a totally different idea: what if instead of overriding `method_missing` and `respond_to_missing`, we override `initialize` on the model so that after the first model has been created, we define the accessors? This avoids preloading the schema and also does not incur any extra performance hit on method calls.

kamipo replied:
Faster idea is welcome, since slower `respond_to?` is sometimes bothoring us.

For example cc2d614, the purpose of cc2d614 is to bypass `respond_to?(:empty?)` in `blank?`, since `respond_to?` on Active Record is much slower due to `DynamicMatchers`.

https://github.com/rails/rails/blob/533dd8a681a3ea7d12acdfc6a2d6a95891282b49/activesupport/lib/active_support/core_ext/object/blank.rb#L18-L20

https://github.com/rails/rails/blob/533dd8a681a3ea7d12acdfc6a2d6a95891282b49/activerecord/lib/active_record/dynamic_matchers.rb#L6-L13

kamipo replied:
Anyway, I agree that it's a relatively narrowly-scoped issue in this PR.

kamipo commented on activemodel/test/cases/validations/acceptance_validation_test.rb:
```diff
@@ -8,7 +8,7 @@
 class AcceptanceValidationTest < ActiveModel::TestCase
   teardown do
-    self.class.send(:remove_const, :TestClass)
+    self.class.send(:remove_const, :TestClass) if self.class.const_defined?(:TestClass)
```
Is this change necessary?

shioyama replied:
Ah sorry no, it was before but I force-pushed a change so it's no longer  required. I'll remove, thanks.

shioyama commented:
Wow noticed there are yet *more* places where things are not being cleaned up after acceptance validators are added to classes:

https://github.com/shioyama/rails/blob/9951f91a12b71f97cab6ea090109a1b37a1ab5d4/activemodel/test/cases/validations/i18n_validation_test.rb#L187-L193

This stuff makes it hard to test changes since totally unrelated specs start breaking... we need to ensure all tests with acceptance validators use subclasses to avoid polluting. Even in case where there is only one module.

shioyama commented:
It just occurred to me that activemodel validations are supposed to work on any class that includes `ActiveModel::Validations`, but the acceptance validator implicitly requires `ActiveModel::AttributeMethods` when it calls `define_attribute_methods`...

kamipo commented:
Calling `define_attribute_methods` is added at ce48b5a, to prevent infinite method_missing loop.
Add `klass.respond_to?(:define_attribute_methods)` is fine to me.

`attribute_method?` is a helper method for validations.

https://github.com/rails/rails/blob/9c87474d82165277a6a5ed3e04b6b88402968134/activemodel/lib/active_model/validations.rb#L270-L272

shioyama commented:
Funny you mention ce48b5a, reverting that change (well, removing that line, I didn't check the git blame) was what got me to this pull request in the first place... I also saw the infinite loop but all else equal, it would be better not to trigger that in a validator override... but now I'm getting stuck.

The basic problem here is that this validator needs to work with both ActiveModel and ActiveRecord. Maybe things would be easier if activerecord just had something custom... otherwise, I think my current solution here is the cleanest one we're going to get.

shioyama commented:
What do you think about just defining an activerecord acceptance validation that subclasses this one?

shioyama commented:
Ok in #35971 I fixed the additional pollution problems mentioned above. I realized that even with the change here, it's still an issue so everything in tests should use subclasses not original test classes.

shioyama commented:
@kamipo how about d44fb37 ? `dup` the module, attach it to the model namespace, then once accessors have been defined, undefine `method_missing` and `respond_to_missing?` from it. That means there is zero performance hit after the first model instance defines any acceptance method.

I think also that this should allow us to remove `define_attribute_methods`, since once the module is removed, there should be no infinite cycle. But since there's no test for that I haven't had a chance to check.

shioyama commented:
Hmm doesn't quite seem to be working just yet...

shioyama commented on activemodel/lib/active_model/validations/acceptance.rb:
```diff
@@ -17,54 +17,61 @@ def validate_each(record, attribute, value)
+
+              def undefine_module_on_class(klass)
+                mod = klass.const_get(:LazilyDefineAttributes)
+                mod.undef_method(:respond_to_missing)
```
This should be `remove_method`, I'll update later.

shioyama commented:
@kamipo This seems to be working now.

Basically:
- we `dup` the original module and name it under the model class, then include it
- anytime we hit `method_missing`, we immediately define acceptance accessors and "implode" the module by removing both `method_missing` and `respond_to_missing?` on it, since we don't need them anymore.

So basically, with this setup:
- only the first instance to hit `method_missing` takes a performance hit on that method, after that the module has no effect (since we've removed the method)
- `respond_to_missing?` will still have a performance hit until we hit `method_missing`, but that should be fixable in the same way (test is failing now on that so haven't committed yet, will investigate)

I'm not an expert on the implications of using `remove_method` but I don't see why this should be an issue. Also, I'm pretty sure this will fix the infinite loop so we shouldn't need that `define_attribute_methods`, but we'll need a test for that first.

What do you think?

shioyama commented:
Ok I've rewritten as a module builder again (as it was originally), but in such a way that there is never more than one. This passes specs for me (some unrelated failures on buildkite, not sure what's up), and should be faster and cleaner than the current solution. It also removes the `define_attribute_methods`, and passes the spec for that which I noticed was added in d99984d48c6.

So I'm happy with this. I realized removing methods on a module is maybe not standard but if it works, why not right? You can't impact performance if you're not there!

shioyama commented:
Some performance results:

```ruby
# frozen_string_literal: true

require "active_record"
require "benchmark/ips"

conn = { adapter: "sqlite3", database: ":memory:" }

ActiveRecord::Base.establish_connection(conn)

class Person < ActiveRecord::Base
  connection.create_table :people, force: :true do |t|
    t.string :name
  end

  validates :terms_of_service, acceptance: true
  validates :eula, acceptance: true
end

class Topic
  include ActiveModel::Validations
  include ActiveModel::Validations::Callbacks
  include ActiveModel::AttributeMethods
  include ActiveSupport::NumberHelper

  validates :terms_of_service, acceptance: true
  validates :eula, acceptance: true

  attr_reader :attributes
  
  def initialize(*)
    @attributes = { 'foo' => nil }
  end

  def attribute(*)
  end
end

Person.create

Benchmark.ips(10) do |x|
  person = Person.new
  foo_person = Person.select('name as foo').first
  topic = Topic.new

  x.report('AR respond_to?') do
    person.respond_to?('miss')
  end

  x.report('AR method_missing') do
    foo_person.foo
  end

  x.report('AM respond_to?') do
    topic.respond_to?('miss')
  end

  x.report('AM method_missing') do
    topic.foo
  end
end
```

**BEFORE**

```
      AR respond_to?    403.265k (± 0.5%) i/s -      4.043M in  10.025001s
   AR method_missing    304.773k (± 0.9%) i/s -      3.055M in  10.024724s
      AM respond_to?    452.244k (± 2.0%) i/s -      4.540M in  10.042971s
   AM method_missing    381.810k (± 1.0%) i/s -      3.852M in  10.090269s
```

**AFTER**

```
      AR respond_to?    865.161k (± 0.5%) i/s -      8.716M in  10.074354s
   AR method_missing    576.058k (± 1.0%) i/s -      5.792M in  10.055779s
      AM respond_to?      1.153M (± 0.9%) i/s -     11.544M in  10.013341s
   AM method_missing    934.458k (± 0.8%) i/s -      9.368M in  10.025936s
```

So with two acceptance attributes, this **more than doubles the performance** of `respond_to?` and `method_missing`. (In the example, I use `select` on the AR model to ensure that there is something hit with `method_missing` (to avoid an exception), and ensure ther is a `foo` key on the AM model for the same reason).

The reason is simple: the changes here remove the overrides after first access on any instance, which happens almost instantaneously, and after that they have zero impact, so the "after" results are basically what you get with no acceptance validation at all.

shioyama commented:
Honestly this is a bigger gain that I had imagined. Basically, any model with an acceptance validator on it right now is getting a pretty massive performance hit on those methods... this just removes that hit.

kamipo commented:
Looks like this opposites to what you ensure in #35893.
Is it acceptable for you?

shioyama commented:
> Looks like this opposites to what you ensure in #35893.

Not really. The issue in #35893 is that the test required method_missing handlers in order to pass, although it does not say that. Also, the situation the test created is one where AR says all attribute metohds have been defined (i.e. `@attribute_methods_generated` is true), but actually they haven't.

The case here where the code calls `define_attribute_methods` here is different (added to avoid infinite loop). And the result is inconsistent and unexpected (I think):

```ruby
class Topic < ActiveRecord::Base
  validates :approved, acceptance: true
end

topic = Topic.new
Topic.undefine_attribute_methods

Topic.instance_variable_get(:@attribute_methods_generated)
#=> false

topic.content # or any column other than "approved"
Topic.instance_variable_get(:@attribute_methods_generated)
#=> false

topic.approved
Topic.instance_variable_get(:@attribute_methods_generated)
#=> true
```

I think `define_attribute_methods` is a very tricky method that should be used with extreme care, and the behaviour above is not what anyone would expect. In both cases I've accessed an attribute method corrresponding to a table column; in one case, this does not redefine methods, in the other it does.

shioyama commented:
I *did* originally want to use a "normal" module rather than an anonymous one. But actually the solution here is I think better because it avoids any arbitrary constants, variables or methods on the model class. Everything is encapsulated and relatively simple, and also performant.

shioyama commented:
I'm going to force-push just to update the comment on f5915ddd47054e8f42b0863c2a263d4b851a0d2f, where I mention we need a test, which I will remove from the commit message. The test was added and the code passes with it.

shioyama commented:
One other thing about `define_attribute_methods`: it is curently in there *exclusively for ActiveRecord models*. For unpersisted models, `define_attribute_methods` with no argument is meaningless: it doesn't define anything.

Which says to me that the activemodel abstraction is broken: stuff is being added that is exclusively for ActiveRecord. Getting rid of that line makes the abstraction slightly less broken.

shioyama commented:
@kamipo Should I add a changelog entry for this?

shioyama commented:
Rebased and added 39f3e2d, which defines the accessor methods on the module rather than on the class itself. I think this makes things cleaner, without changing behaviour.

shioyama commented:
@kamipo Is there anything else I should do for this one? It is passing tests (plus some new ones I added) and it improves performance a lot + fixes memory leak issue.

shioyama commented:
@kamipo This PR seems to have gone quiet. I've just rebased it now so it's up to master. I think it's a good PR with no issues AFAICT, and benefits including a performance boost to `respond_to?` and cleaner code without memory leak issues. What do you think?

rafaelfranca commented:
@shioyama thank you for the PR. It looks good. Nice improvement. Can you rebase it?

kamipo commented:
> Looks like this opposites to what you ensure in #35893.
Is it acceptable for you?

What I mean is that we have a test for extra definining attributes after attribute methods are defined.

https://github.com/rails/rails/blob/41bc4c6207147a5eeafa323d763e66c06e61cacf/activerecord/test/cases/attributes_test.rb#L233-L235

In the current implementation in this PR, extra definining acceptance attributes will not be worked.

shioyama commented:
> In the current implementation in this PR, extra definining acceptance attributes will not be worked.

I could just leave the `method_missing`/`respond_to?` methods in-place, but then we lose the performance benefit.

shioyama commented:
I wonder if it makes sense to deprecate the (undocumented?) feature of being able to add attributes to a model class after it has instances created. This seems to be a pretty narrow edge-case which makes it hard/impossible to optimize in cases like this one.

shioyama commented:
@rafaelfranca This is rebased and passing :smile:

kamipo commented:
I'm not fond changing behavior without noticing.

If the target scope of this PR is only fixing #34659, then I can approve.

But if we intentionally change current behavior for any benefits, I think we should have test what is changed (e.g. https://github.com/rails/rails/pull/36052/files#diff-d80af68291f2bb61c9fccc01482df877R211).

shioyama commented:
@kamipo I added 9176654 which addresses the case where a new acceptance validation is defined after the module has already been "triggered" via method_missing. I also added a test checking this behaviour.

Basically I did this by defining `LazilyDefineAttributes===` so that it excludes modules which have had their methods removed. Thus if you define a new acceptance validation, it will not see the previous one (with methods removed) and will include it again, allowing it to trigger for the new attribute(s).

Concretely speaking, this is the case where this is needed:

```ruby
class Topic
  validates_acceptance_of(:something_to_accept)
end

topic = klass.new
topic.something_to_accept

Topic.validates_acceptance_of(:something_else_to_accept)

topic.something_else_to_accept
```

Previously, the first call to `something_to_accept` would define accessors and remove the module, so you wouldn't be able to define any other acceptance validations. But now it also marks the module as "removed" so if you define a validator again, it will include a new module and `something_else_to_accept` will be caught by the new module's `method_missing`.

Make sense?

shioyama commented:
@rafaelfranca @kamipo Any update on this one? I've addressed @kamipo's [concern](https://github.com/rails/rails/pull/35969#issuecomment-515311664), the change improves performance of `respond_to?` and `method_missing`, and avoids multiple module inclusion (as originally mentioned in the title).

If there's anything else to address, I'm happy to do that. I can also squash commits if necessary.

kamipo commented:
Yes, can you squash your commits?

shioyama commented:
Rebased and squashed! Seem to be some unrelated failing specs though.
